### ISSUE TITLE
IDL Program

### ISSUE THREAD

**Issue by armaniferrante** [solana-foundation/anchor#1180]
There should be a separate program that stores the IDLs as PDAs of a program id and version number, instead of embedding the IDL instructions into the program directly.

**Comment by kevinaud** [comment]
Adding the diagrams from the discord thread for posterity.

Go from this:
![Anchor Modules - Bundled approach](https://user-images.githubusercontent.com/13792949/146798157-d894c8e2-f2cd-440e-a47a-dcc0a8f04764.png)

To this:
![Anchor Modules - Generic approach](https://user-images.githubusercontent.com/13792949/146798200-3ecbd576-1779-4ab0-bbe6-600921761468.png)

**Comment by armaniferrante** [comment]
An additional question we should consider is how to atomically update the program binary with the IDL. This is important for explorers. There are two options

1. Add a custom ELF section (can you do that?) to the .so file with the IDL, or append the IDL to the end of the bytecode.
2. Assume the user uses the Anchor CLI to deploy and provide a new command to do the program upgrade/deploy + IDL upgrade/deploy in the same tx.

**Comment by kevinaud** [comment]
I'm not sure if I fully grasp #1, I'm not familiar with the details of solana executables (and I got a C in operating systems haha) but #2 sounds reasonable to me.

A question I have is how do we manage the ownership/deployment of the IDL Program? Someone (I'm assuming you) will have to generate a keypair and be responsible for keeping the private key safe + deploying the program. We could set up a CI/CD deployment pipeline and store the encrypted private key so it can auto-deploy.

**Comment by kevinaud** [comment]
Someone on discord suggested using this project for key management/deployment: https://github.com/GokiProtocol/goki

I don't have experience with it but it looks interesting.

**Comment by armaniferrante** [comment]
Dealing with upgrading the IDL program is probably the main argument against such a program. We would need some governance (probably a DAO) around it. My instinct is that it's better to just leave that up to the governance mechanism of the associated program rather than introduce another one.

**Comment by kevinaud** [comment]
I agree that is probably the main argument against this. Do you have any idea how the SPL handles this for those programs?

**Comment by paul-schaaf** [comment]
IDL generation is an anchor feature and the SPL programs dont use anchor so they dont have to think about it

**Comment by paul-schaaf** [comment]
I'm now also of the opinion that it would be better to keep it inside the programs for the reason that I think introducing any kind of dao or similiar governance structure around anchor will slow us down too much.

In the future, this could be revisited tho imo. Having a separate program has the benefit that upgrading the IDL of all anchor programs, e.g. in case there's a bug in it, can be done through upgrading a single program.

Without an IDL program, if there's a bug in the IDL generation code, each anchor program has to upgrade themselves. For some this may be easy, for others, whose upgrades are subject to governance and timelocks, it will be more difficult. In addition, upgrading to a new anchor version would also mean incorporating other new anchor code into their projects (how much of a bad thing that it, I'm not sure yet).

**Comment by kevinaud** [comment]
I know SPL programs don't use IDL, I just meant how do they handle governance/upgrades

**Comment by jacobcreech** [comment]
We've come full circle now with https://github.com/solana-foundation/anchor/issues/3644. Pretty cool to see how far ahead you were Armani. Closing for now with #3644 being the tracker for implementation.
